{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "4fe4ddb1",
   "metadata": {},
   "source": [
    "# GENE REPRESENTATIONS FOR DEEP GWAS"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "21697f76",
   "metadata": {},
   "source": [
    "## Workspace Setup"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "87800b4a",
   "metadata": {},
   "source": [
    "### Packages"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "f3505419",
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import pandas as pd\n",
    "import scipy as sp\n",
    "import torch as pt"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a0ac0df6",
   "metadata": {},
   "source": [
    "### Global Variables"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d63525cf",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "57afb50f",
   "metadata": {},
   "source": [
    "### Functions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "b302bdae",
   "metadata": {
    "code_folding": [
     0,
     51,
     93,
     141,
     189,
     223
    ]
   },
   "outputs": [],
   "source": [
    "def create_samples(n,nc,p,l,m,i):\n",
    "    '''Creates a set of artificial-genotype samples for one single gene across multiple samples of a population.\n",
    "    \n",
    "    parameters\n",
    "    ----------\n",
    "    n: `int`\n",
    "        number of samples\n",
    "    nc: `int`\n",
    "        number of causal SNPs\n",
    "    p: `float`\n",
    "        prevalence (i.e., minor-class ratio) in [0,1]\n",
    "    l: `int`\n",
    "        gene length in terms of SNPs (rather than bp),\n",
    "    m: `int`\n",
    "        maximum number of (minor-allele) SNPs present in a sample\n",
    "    i: `bool`\n",
    "        whether or not SNP interaction is required to produce the minor-class phenotype\n",
    "        \n",
    "    returns\n",
    "    -------\n",
    "    X: `set`\n",
    "        set samples w/ desired prevalence of minor class\n",
    "    C: `set`\n",
    "        set of causal-SNP identifiers\n",
    "    '''\n",
    "    # create a the set of SNP identifiers\n",
    "    S = list(np.arange(1,l+1))\n",
    "    \n",
    "    # random choice of causal SNPs\n",
    "    C = list(np.random.choice(S,size=nc,replace=False))\n",
    "    \n",
    "    # define set of non-causal SNPs\n",
    "    D = [s for s in S if s not in C]\n",
    "    \n",
    "    # calculate number of positive/negative samples required\n",
    "    n_pos = int(p*n)\n",
    "    n_neg = n - n_pos\n",
    "    \n",
    "    # create positive and negative samples w/ or w/o interaction\n",
    "    if i:\n",
    "        P = pos_samples_and(n_pos,m,C,D)\n",
    "        N = neg_samples_and(n_neg,m,C,D)\n",
    "    else:\n",
    "        P = pos_samples_or(n_pos,m,C,D)\n",
    "        N = neg_samples_or(n_neg,m,D)\n",
    "        \n",
    "    # create entire set of samples as union of positive and negative samples\n",
    "    X = P+N\n",
    "    \n",
    "    return X,C\n",
    "\n",
    "def pos_samples_and(n_pos,m,C,D):\n",
    "    '''Creates a set of minor-class artificial-genotype samples for one single gene under SNP interaction.\n",
    "    \n",
    "    parameters\n",
    "    ----------\n",
    "    n_pos: `int`\n",
    "        number of positive samples\n",
    "    m: `int`\n",
    "        maximum number of SNPs present\n",
    "    C: `set`\n",
    "        set of causal SNPs\n",
    "    D: `set`\n",
    "        set of non-causal SNPs\n",
    "        \n",
    "    returns\n",
    "    -------\n",
    "    P: `set`\n",
    "        set minor-class samples\n",
    "    '''\n",
    "    # determine number of causal SNPs\n",
    "    nc = len(C)\n",
    "    \n",
    "    # initialize P as empty set\n",
    "    P = []\n",
    "    \n",
    "    # loop over minor-class samples to be created\n",
    "    for k in range(n_pos):\n",
    "        \n",
    "        # random choice of number of non-causal SNPs present\n",
    "        nk = np.random.randint(low=1,high=m-nc+1)\n",
    "        \n",
    "        # random choice of non-causal SNPs present\n",
    "        Dk = list(np.random.choice(D,size=nk,replace=False))\n",
    "        \n",
    "        # create sample as union of all causal and chosen non-causal SNPs\n",
    "        Pk = C+Dk\n",
    "        \n",
    "        # add sample to list of samples\n",
    "        P = P+[Pk]\n",
    "    \n",
    "    return P\n",
    "\n",
    "def neg_samples_and(n_neg,m,C,D):\n",
    "    '''Creates a set of major-class artificial-genotype samples for one single gene under SNP interaction.\n",
    "    \n",
    "    parameters\n",
    "    ----------\n",
    "    n_neg: `int`\n",
    "        number of negative samples\n",
    "    m: `int`\n",
    "        maximum number of SNPs present\n",
    "    C: `set`\n",
    "        set of causal SNPs\n",
    "    D: `set`\n",
    "        set of non-causal SNPs\n",
    "        \n",
    "    returns\n",
    "    -------\n",
    "    N: `set`\n",
    "        set major-class samples\n",
    "    '''\n",
    "    # determine number of causal SNPs\n",
    "    nc = len(C)\n",
    "    \n",
    "    # random choice of number of causal SNPs present (maximally all but one)\n",
    "    s = np.random.randint(low=0,high=nc,size=n_neg)\n",
    "    \n",
    "    # initialize N as empty set\n",
    "    N = []\n",
    "    \n",
    "    # loop over minor-class samples to be created\n",
    "    for k in range(n_neg):\n",
    "        \n",
    "        # random choice of number of non-causal SNPs present\n",
    "        nk = np.random.randint(low=1,high=m-s[k]+1)\n",
    "        \n",
    "        # random choice of causal SNPs present\n",
    "        Ck = list(np.random.choice(C,size=s[k],replace=False))\n",
    "        \n",
    "        # random choice of non-causal SNPs present\n",
    "        Dk = list(np.random.choice(D,size=nk,replace=False))\n",
    "        \n",
    "        # create sample as union of all causal and chosen non-causal SNPs\n",
    "        Nk = Ck+Dk\n",
    "        \n",
    "        # add sample to list of samples\n",
    "        N = N+[Nk]\n",
    "    \n",
    "    return N\n",
    "\n",
    "def pos_samples_or(n_pos,m,C,D):\n",
    "    '''Creates a set of minor-class artificial-genotype samples for one single gene w/o SNP interaction.\n",
    "    \n",
    "    parameters\n",
    "    ----------\n",
    "    n_pos: `int`\n",
    "        number of positive samples\n",
    "    m: `int`\n",
    "        maximum number of SNPs present\n",
    "    C: `set`\n",
    "        set of causal SNPs\n",
    "    D: `set`\n",
    "        set of non-causal SNPs\n",
    "        \n",
    "    returns\n",
    "    -------\n",
    "    P: `set`\n",
    "        set minor-class samples\n",
    "    '''\n",
    "    # determine number of causal SNPs\n",
    "    nc = len(C)\n",
    "    \n",
    "    # random choice of number of causal SNPs present (minimally one)\n",
    "    s = np.random.randint(low=1,high=nc+1,size=n_pos)\n",
    "    \n",
    "    # initialize P as empty set\n",
    "    P = []\n",
    "    \n",
    "    # loop over minor-class samples to be created\n",
    "    for k in range(n_pos):\n",
    "        \n",
    "        # random choice of number of non-causal SNPs present\n",
    "        nk = np.random.randint(low=1,high=m-s[k]+1)\n",
    "        \n",
    "        # random choice of causal SNPs present\n",
    "        Ck = list(np.random.choice(C,size=s[k],replace=False))\n",
    "        \n",
    "        # random choice of non-causal SNPs present\n",
    "        Dk = list(np.random.choice(D,size=nk,replace=False))\n",
    "        \n",
    "        # create sample as union of all causal and chosen non-causal SNPs\n",
    "        Pk = Ck+Dk\n",
    "        \n",
    "        # add sample to list of samples\n",
    "        P = P+[Pk]\n",
    "    \n",
    "    return P\n",
    "\n",
    "def neg_samples_or(n_neg,m,D):\n",
    "    '''Creates a set of major-class artificial-genotype samples for one single gene w/o SNP interaction.\n",
    "    \n",
    "    parameters\n",
    "    ----------\n",
    "    n_neg: `int`\n",
    "        number of negative samples\n",
    "    m: `int`\n",
    "        maximum number of SNPs present\n",
    "    D: `set`\n",
    "        set of non-causal SNPs\n",
    "        \n",
    "    returns\n",
    "    -------\n",
    "    N: `set`\n",
    "        set major-class samples\n",
    "    '''\n",
    "    # random choice of number of non-causal SNPs present\n",
    "    s = np.random.randint(low=1,high=m+1,size=n_neg)\n",
    "    \n",
    "    # initialize N as empty set\n",
    "    N = []\n",
    "    \n",
    "    # loop over minor-class samples to be created\n",
    "    for k in range(n_neg):\n",
    "        \n",
    "        # random choice of non-causal SNPs present\n",
    "        Nk = list(np.random.choice(D,size=s[k],replace=False))\n",
    "        \n",
    "        # add sample to list of samples\n",
    "        N = N+[Nk]\n",
    "    \n",
    "    return N\n",
    "\n",
    "def assign_phenotype(s,C,i):\n",
    "    '''Assigns class label based on a single-gene sample and ground-truth of causal SNPs and interaction.\n",
    "    \n",
    "    parameters\n",
    "    ----------\n",
    "    s: `set`\n",
    "        set of present (minor-allele) SNPs\n",
    "    C: `set`\n",
    "        set of causal-SNP identifiers\n",
    "    i: `bool`\n",
    "        whether or not SNP interaction is required to produce the minor-class phenotype\n",
    "        \n",
    "    returns\n",
    "    -------\n",
    "    p: `int`\n",
    "        binary phenotype, 1/0 for minor/major class\n",
    "    '''\n",
    "    # create list of causal SNPs present\n",
    "    l = [c for c in s if c in C]\n",
    "    \n",
    "    # in case of required SNP interaction\n",
    "    if i:\n",
    "        \n",
    "        # if all causal SNPs are present \n",
    "        if len(l)==len(C):\n",
    "            # sample is positive\n",
    "            p = 1\n",
    "        else:\n",
    "            p = 0\n",
    "            \n",
    "    # in case of no SNP inetraction    \n",
    "    else:\n",
    "        \n",
    "        # if at least one causal SNPs present \n",
    "        if len(l)>=1:\n",
    "            # sample is positive\n",
    "            p = 1\n",
    "        else:\n",
    "            p = 0\n",
    "            \n",
    "    return p"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "715aef49",
   "metadata": {},
   "source": [
    "## Data Generation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "id": "9e0efd8e",
   "metadata": {},
   "outputs": [],
   "source": [
    "n  = 100\n",
    "nc = 3\n",
    "p  = .35\n",
    "l  = 20\n",
    "m  = 20-nc\n",
    "i  = False\n",
    "\n",
    "# create sample dictionary\n",
    "X,C = create_samples(n,nc,p,l,m,i)\n",
    "\n",
    "sample_dict = {}\n",
    "for j,s in enumerate(X):\n",
    "    sample_dict[j+1] = s\n",
    "    \n",
    "# label samples\n",
    "label_dict = {}\n",
    "for s in sample_dict:\n",
    "    label_dict[s] = assign_phenotype(sample_dict[s],C,i)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "01f472b0",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.5"
  },
  "latex_envs": {
   "LaTeX_envs_menu_present": true,
   "autoclose": false,
   "autocomplete": true,
   "bibliofile": "biblio.bib",
   "cite_by": "apalike",
   "current_citInitial": 1,
   "eqLabelWithNumbers": true,
   "eqNumInitial": 1,
   "hotkeys": {
    "equation": "Ctrl-E",
    "itemize": "Ctrl-I"
   },
   "labels_anchors": false,
   "latex_user_defs": false,
   "report_style_numbering": false,
   "user_envs_cfg": false
  },
  "toc": {
   "base_numbering": 1,
   "nav_menu": {},
   "number_sections": true,
   "sideBar": true,
   "skip_h1_title": true,
   "title_cell": "Table of Contents",
   "title_sidebar": "Contents",
   "toc_cell": false,
   "toc_position": {},
   "toc_section_display": true,
   "toc_window_display": false
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
